### DISCUSSION TITLE
proposals combination

### DISCUSSION THREAD

**Discussion by RabeaAfifi** [strongswan/strongswan#1224] [Q&A] [answered]
Hi 
I want to set proposal options for possible ipsec conection 
based on the IKEv2 cipher suit doc page on here
https://docs.strongswan.org/docs/5.9/config/IKEv2CipherSuites.html
I have picked these values

for Authenticated Encryption (AEAD) Algorithms :
I choose aes256gcm16, aes192gcm16, aes128gcm16, chacha20poly1305, aes256ccm16

and for Integrity Algorithms :
sha2_512,s ha2_384, sha2_256, aescmac

and for PRF I choose : 
prfsha512 , prfsha384 , prfsha256 , prfaesxcbc

and for Diffie Hellman Groups I choose:
ecp512bp, ecp521, ecp384bp

so for the configuration on in swanctl.conf ,I need to add a connection with 
proposal = aes256gcm16-sha2_512-prfsha512-ecp512bp , aes92gcm16-sha2_512-prfsha512-ecp512bp .....

so the proposals with be very long and very explicit, is there a way to select possible options for each part in the proposal separately and let strongswan do the combinations for me ?
What if the options get up 1000 in all the combinations , do I really have to type all 1000 proposal options separated with comma ?

**Comment by strongX509** [comment, marked as answer]
Just write
````
proposals = aes256gcm16-aes192gcm16-aes128gcm16-chacha20poly1305-aes256ccm16-prfsha512-prfsha384-prfsha256-prfaesxcbc-ecp512bp-ecp521-ecp384bp
````
and you get all combinations.

**Reply by RabeaAfifi** [reply]
Thanks a lot , it worked.
and I was wondering , can I put the prf section as optional in the same line by using some special character 
instead of creating the same line but without the prfs,
also I am searching in the documentation and did not find the syntax options for proposals string, can you help with it ?

**Reply by strongX509** [reply]
What is the intention? Re-using the IKE `proposals` string for `esp_proposals`? This doesn't work. You have to define two distinct strings.

The `proposals` structure is explained in https://docs.strongswan.org/docs/5.9/swanctl/swanctlConf.html:

> Algorithm keywords get separated using dashes. Multiple proposals may be separated by commas.
